test/e2e/source_ping_test.go

```diff
@@ -108,8 +119,9 @@ func TestPingSourceV1Alpha2(t *testing.T) {
 	client.WaitForAllTestResourcesReadyOrFail()
 
 	// verify the logger service receives the event and only once
-	if err := client.CheckLog(loggerPodName, lib.CheckerContainsCount(data, 1)); err != nil {
-		t.Fatalf("String %q not found or found multiple times in logs of logger pod %q: %v", data, loggerPodName, err)
+	err = targetTracker.WaitMatchSourceData(sourcesv1alpha2.PingSourceSource(client.Namespace, sourceName), data, 1, 1)
```

Comment:
Ok a little braindump here: so assertions are in the form of "wait for that data" or "wait for attribute x" right? Can we have some "list" like api, like `GetAllReceivedEvents()` or `GetEvent(eventid)` that blocks until i have eventid? So we can write more "straightforward" tests like:

* set up the components
* send event with id X
* receive event with id X
* run asserts on event with id X

Reply:
Yeah, it's definitely possible to do it that way as well.   In cases where we have an eventid, that works nicely.  If you call Find() with a match all it will actually return the list of matching structures.  You still have to figure out how to wait for events to show up, etc.  One nice thing about not doing the GetEvent() logic externally is I have some logic in the main Wait* function that these are using that if it times out without seeing sufficient matching events, it prints the 5 most recently seen events(even if they didn't match the pattern) in the test code.  It makes it MUCH easier to figure out what's going on in cases where the filter doesn't match the event.  We don't want to have to reimplement that logic everywhere.

Reply:
ah ok so `Find()` does the polling until the predicate matches right? Because that's what I'm asking for: `GetEvent(eventId)` as explained by me on the first message is a blocking method that waits for the event with the specified id.

So does `Find()` behaves like you explained (prints the last 5 messages etc)? Because in that case it rocks!

Reply:
Sorry, for some reason I misread.  You Find() just finds what is currently there and. returns them all.  GetEvent(eventId) would be a trivial wrapper around the normal WaitAtLeastNMatch() call though.  The one issue for some of the tests is we'll have to be better about making sure all components are fully up in the right order than i think we probably are today.  IE, some of the tests today just use regular heartbeat type data, so they can get away with ignoring that assuming that as long as everything is up, they will eventually work.

Reply:
I'm not sure i fully understand this problem

Reply:
So `WaitAtLeastNMatch()` waits for N events matching the predicate right? because in that case i think it's already enough for the flow explained in the first comment